pkg/nvidia/nvfanctrl/main.go

```diff
@@ -0,0 +1,366 @@
+	if errF != nil {
+		return fmt.Errorf("Fail to change FAN speed")
+	}
+	return nil
+}
+
+// Perform the controlling iteration
+func (pwm fanCTRL) controlPWM() error {
+	temp, errT := pwm.readTemp()
+	if errT != nil {
+		return fmt.Errorf("Fail to read temperature!")
+	}
+
+	pwmValue := maxPWM // Set to maximum
+	for _, v := range pwm.cmode.points {
+		if temp < v[0] {
+			pwmValue = v[1]
+			break
+		}
+	}
+
+	return pwm.setPWM(pwmValue)
+}
+
+// Controlling loop
+func (pwm fanCTRL) run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-time.Tick(time.Duration(pwm.checkInterval) * time.Second):
+			err := pwm.controlPWM()
+			if err != nil {
+				fmt.Println("Error:", err)
+			}
+		}
+	}
+}
+
+// Stop FAN controller and set PWM to maximum
+func (pwm fanCTRL) finish() error {
+	fmt.Println("Finishing FAN controller")
+	return pwm.setPWM(maxPWM)
+}
+
+func main() {
+	var mode string
+	var tzone int
+	var checkInterval int
+	var cmode coolingMode
+
+	// Build and validate the command line
+	flag.Usage = func() {
+		fmt.Printf("FAN PWM controller for Jetson devices\n\n")
+		fmt.Printf("Use:\n    %s [-m <mode>] [-t <thermal_zone>] [-i <pooling_time>]\n\n", os.Args[0])
+		flag.PrintDefaults()
+	}
+	flag.StringVar(&mode, "m", "quiet", "Cooling profile: quiet or cool")
+	flag.IntVar(&tzone, "t", 0, "Thermal zone number")
+	flag.IntVar(&checkInterval, "i", 2, "Pooling time (in seconds)")
+	flag.Parse()
+
+	// Identify device and get PWM profiles
+	dev, err := getDeviceProfile()
+	if err != nil {
+		dev = profileXavier{}
+		fmt.Fprintf(os.Stderr, "Device could not be identified, using %s profile\n", dev.Model())
+	}
+
+	switch (mode) {
+	case "quiet":
+		cmode = dev.getQuietProfile()
+	case "cool":
+		cmode = dev.getCoolProfile()
+	default:
+		fmt.Fprintf(os.Stderr, "Invalid cooling profile: %s\n", mode)
+		os.Exit(1)
+	}
+	if !checkThermalZone(tzone) {
+		fmt.Fprintf(os.Stderr, "Invalid thermal zone: %d\n", tzone)
+		os.Exit(1)
+	}
+	if checkInterval <= 0 || checkInterval >= 60 {
+		fmt.Fprintf(os.Stderr, "Invalid pooling time: %d\n", checkInterval)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Starting FAN controller (device identified: %s)\n", dev.Model())
+
+	// Search the PWM fan device
+	fanDevice, err := getFANDevice()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Could not found a FAN device to control\n")
+		os.Exit(1)
+	}
+
+	// FAN controller
+	pwmCTRL := fanCTRL{
+		devPWM: fanDevice,
+		devThermal: getThermalDevice(tzone),
+		cmode: cmode,
+		checkInterval: checkInterval,
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGHUP)
+	defer func(c chan<- os.Signal) {
+		signal.Stop(c)
+		cancel()
+	}(sigChan)
+
+	go func() {
+		for {
+			select {
+			case s := <-sigChan:
+				if (s == os.Interrupt) {
+					cancel()
+					if err := pwmCTRL.finish(); err != nil {
```

Comment:
I didn't get your rationale @OhmSpectator , I need to catch interrupt signals in order to ensure we do the proper exit (in this case, set the FAN PWM to maximum), the execution from main is completely interrupted once the interrupt signals arrive, that's why I have a thread that is going to be blocked until we receive the signals, and it will handle them performing the gracefully exit....

Reply:
We discussed it in DM. Let it be this way here, it does not hurt.